### PR TITLE
gh-126384: Add tests to verify the behavior of basic COM methods.

### PR DESCRIPTION
**Pull request by junkmd**
I would like to backport this to 3.12 and 3.13 as well.
This is internal-only, so I don’t think it needs a NEWS entry.

* Issue: gh-126384

### CODE REVIEW

**Comment by junkmd** [comment]
I renamed utility functions for testing to conform with PEP8. 

Also I added URLs in comments to documentation on COM specifications for developers who may have limited experience with COM and may review or maintain this code in the future.

File: Lib/test/test_ctypes/test_win32_com_foreign_func.py
Comment (encukou) [inline Lib/test/test_ctypes/test_win32_com_foreign_func.py:7]: The whole file is Windows-specific, so it would make sense to skip it here:
```suggestion

if sys.platform != "win32":
    raise unittest.SkipTest("Windows-specific test")
```
After this line you can import `HRESULT` or `ole32`, create the GUIDs, etc.

Comment (junkmd) [reply Lib/test/test_ctypes/test_win32_com_foreign_func.py:7]: Good catch.

As you pointed out, writing the platform bridge directly at the module level, rather than inside `setUpModule` as I did in the earlier commits, would indeed ensure that the tests are appropriately skipped on non-Windows platforms.

File: Lib/test/test_ctypes/test_win32_com_foreign_func.py
Comment (encukou) [inline Lib/test/test_ctypes/test_win32_com_foreign_func.py]: This is a very convoluted way of collecting the necessary arguments. IMO, it distracts from the point of the test. As far as I can see, `ProtoComMethod` can be replaced with:
```py
def ProtoComMethod(name, index, restype, *argtypes):
    proto = ctypes.WINFUNCTYPE(restype, *argtypes)
    def make_method(*args):
        foreign_func = proto(index, name, *args)
        def call(self, *args, **kwargs):
            return foreign_func(self, *args, **kwargs)
        return call
    return make_method
```
with the difference in usage being that `name` needs to be passed in explicitly, as in:
```diff
     proto_query_interface = ProtoComMethod(
+       'QueryInterface',
         0, HRESULT, POINTER(GUID), POINTER(c_void_p)
     )
```
This way the `instancemethod` type (`PyInstanceMethod_New` and `PyInstanceMethod_Type` above) isn't needed.

**Review by encukou — COMMENTED** [review]
Thank you for the tests!
I only have a slight idea of how COM works, but, they look good to me!

However, I would like to simplify the supporting Python code, though: we don't need to test CPython internal types, or custom descriptors using `__get_name__`.

File: Lib/test/test_ctypes/test_win32_com_foreign_func.py
Comment (junkmd) [reply Lib/test/test_ctypes/test_win32_com_foreign_func.py]: I thought it was a cool idea to use the descriptor protocol to avoid having to specify a string with the same name as the attribute.  
But you're right—this is test code, not the package's production code. As you pointed out, even if it's redundant, keeping it as simple as possible is better.

Comment (junkmd) [reply Lib/test/test_ctypes/test_win32_com_foreign_func.py]: > This way the `instancemethod` type (`PyInstanceMethod_New` and `PyInstanceMethod_Type` above) isn't needed.

To be honest, I didn’t really understand why `instancemethod` was necessary in the previous implementation.
I just reused the implementation used in `comtypes`.

**Comment by junkmd** [comment]
In #126686, `COMError` became public, but I would like this test to be backported to 3.13 and 3.12, where `COMError` remains private.  

If possible, I would appreciate it if we could merge this as is, and after the backport PR is submitted by the bot, I will prepare a follow-up PR reflecting the fact that `COMError` is now public.

File: Lib/test/test_ctypes/test_win32_com_foreign_func.py
Comment (picnixz) [inline Lib/test/test_ctypes/test_win32_com_foreign_func.py:8]: Another way to do it is to raise if the import of COMError fails.

**Review by picnixz — COMMENTED** [review]
Can we have assertion equality being tested as actual == expected instead of expected == actual? I think we usually do the first form rather than the second one.

File: Lib/test/test_ctypes/test_win32_com_foreign_func.py
Comment (encukou) [reply Lib/test/test_ctypes/test_win32_com_foreign_func.py:8]: I'd not do that. If `COMError` fails to import on Windows, it's a bug.

File: Lib/test/test_ctypes/test_win32_com_foreign_func.py
Comment (encukou) [inline Lib/test/test_ctypes/test_win32_com_foreign_func.py:43]: For my previous suggestion, I stopped simplifying too early:
```suggestion
        return proto(index, name, *args)
```

**Review by encukou — APPROVED** [review]
Looks good, thank you!

I don't think it's worth it to reorder assertion arguments, but if you want to do it I won't stop you :)

**Comment by encukou** [comment]
[Edit: I deleted a misguided suggestion]

**Comment by junkmd** [comment]
I don’t have a strong preference for the order of `actual == expected` or `expected == actual`.
If it’s mergeable as is, please go ahead and merge it.

In my native language, Japanese, the equivalent concept is expressed as “予実” (yojitsu), with the characters for “expected” and “actual” written in that order.
That might be why I wrote it in this order.

**Review by picnixz — APPROVED** [review]
Let's not burden ourselves with an additional commit. I was just surprised to see it written like that but I won't die from it :) (and since it's a new file anyway, it can have its own style).

By the way, do we need a `if __name__ == '__main__'` block executing the unittest as is?

**Comment by junkmd** [comment]
I added the block.

**Comment by picnixz** [comment]
Ah thank you. I wasn't sure whether it was needed or not. But one build bot is failing I think?

**Comment by picnixz** [comment]
!buildbot Windows

**Comment by junkmd** [comment]
I misunderstood the type of bot running in CI.
I thought all the bots specified with `!buildbot Windows` were running in CI.

I’m not sure why the bot is failing.
There are no changes to production code this time.
Looking at the bot’s traceback, it seems that something related to f-strings isn’t working correctly, but the test code I added this time doesn’t use any f-strings.

**Comment by junkmd** [comment]
!buildbot Windows

**Comment by encukou** [comment]
Sorry, I had to step away yesterday.

> By the way, do we need a if __name__ == '__main__' block executing the unittest as is?

No, that's not necessary in new tests. But it doesn't hurt.

> I misunderstood the type of bot running in CI.

The Buildbot tests run *after* merge, as they can be slow or unreliable. And they run on volunteer-maintained hardware; for security we only run them on reviewed code.
I triggered them in order to check the test on several versions of Windows, e.g. 32-bit or ARM.
Sometimes they find issues in uncommon setups. Usually, some fail for unrelated reasons; those can be ignored. (GitHub unfortunately shows a scary-looking red message if any reported tests fail.)

**Pull request merged**
